_posts/2024-07-01-zio-fullstack-webapp.md

```diff
@@ -0,0 +1,844 @@
+
+
+Quill is a compile-time query language for Scala and SQL. It allows to write queries in Scala and compile them to SQL.
+
+
+Quill compiles **(4a)** and **(4b)** to SQL like this:
+
+
+```sql
+INSERT INTO users (name,email,age,creation_date) VALUES (?, ?, ?, ?) RETURNING id, name, email, age, creation_date;
+SELECT x4.id, x4.name, x4.email, x4.age, x4.creation_date AS creationDate FROM users x4 WHERE x4.id IS NULL AND ? IS NULL OR x4.id = ?;
+```
+
+
+### 5.1.3. Flyway Service
+
+In the same way, Flyway service is a layer that will handle the database migrations.
+
+```scala
+object FlywayServiceLive {
+  def live: RLayer[FlywayConfig, FlywayService] = ZLayer( // (1)
+    for {
+      config <- ZIO.service[FlywayConfig]
+      flyway <- ZIO.attempt(Flyway.configure()
+                   .dataSource(config.url, config.user, config.password).load())
+    } yield new FlywayServiceLive(flyway)
+  )
+
+  val configuredLayer: TaskLayer[FlywayService] = Configs.makeConfigLayer[FlywayConfig]("db.dataSource") 
+                                                             >>> live  // (2)
+}
+```
+  
+* **(1)** is a layer that will provide a **FlywayService**, it depends on the availability of **FlywayConfig** service.
+* **(2)** Inject a **FlywayConfig** in a the previous layer and the remove the dependency on the config.
+
+### 5.1.4. Person Service
+
+Person service is a layer that will handle the business logic of the application.
+
+```scala
+trait PersonService { // Service trait
+  def register(person: Person): Task[User]
+}
+                                  // Dependency injection
+class PersonServiceLive private (userRepository: UserRepository) extends PersonService {
+  def register(person: Person): Task[User] =
+    userRepository.create(
+      User(
+        id = None,
+        name = person.name,
+        email = person.email,
+        age = person.age,
+        creationDate = ZonedDateTime.now()
+      )
+    )
+}
+
+object PersonServiceLive {           // Layer derivation (macro) from the constructor
+  val layer: RLayer[UserRepository, PersonService] = ZLayer.derive[PersonServiceLive]
+}
+```
+
+
+### 5.1.5. Server layer
+
+Server layer is a layer that will handle the HTTP server, is provided by ZIO Http with convenient default configuration.
+
+It is a dependency that will be provided to the server effect.
+
+```scala
+private val server: ZIO[PersonService & Server, IOException, Unit] =
+    for {
+      _         <- Console.printLine("Starting server...")
+      endpoints <- HttpApi.endpointsZIO                // (1)
+      docEndpoints = SwaggerInterpreter()              // (2)
+                       .fromServerEndpoints(endpoints, "zio-laminar-demo", "1.0.0")
+      _ <- Server.serve(                               // (3)
+             ZioHttpInterpreter(serverOptions)
+               .toHttp(metricsEndpoint :: webJarRoutes :: endpoints ::: docEndpoints)
```

Comment:
Can you explain this composition of endpoints? Even if you're building a simple list, it's worth explaining because the operators are not immediately understandable